_blog/2024-02-08-uptime-monitoring-rust.mdx

```diff
@@ -0,0 +1,791 @@
+
+        let number_of_seconds = match split_by {
+            SplitBy::Hour => 3600,
+            SplitBy::Day => 86400,
+        };
+```
+
+Then using the previous variables, we can dynamically fill in missing data points and return `data`!
+
+```rust
+if data.len() < number_of_splits {
+    for i in 1..number_of_splits {
+        let created_at = Utc::now().format(date_format).to_string();
+        let created_at =
+            DateTime::parse_from_str(&created_at, "%Y/%m/%d %H:%M:%S%.3f %z").unwrap() -
+            chrono::Duration::seconds((number_of_seconds * i).try_into().unwrap());
+
+        if !data.iter().any(|x| x.time == created_at) {
+            data.push(WebsiteStats {
+                time: created_at.into(),
+                uptime_pct: None,
+            });
+        }
+    }
+    data.sort_by(|a, b| b.time.cmp(&a.time));
+}
+
+data
+```
+
+Now we can refactor our handler function:
+
+```rust
+async fn get_websites(State(state): State<AppState>) -> impl AskamaIntoResponse {
+    let websites = sqlx
+        ::query_as::<_, Website>("SELECT url, alias FROM websites")
+        .fetch_all(&state.db).await
+        .unwrap();
+
+    let mut logs = Vec::new();
+
+    for website in websites {
+        let data = get_website_uptime_stats(SplitBy::Hour, &website.alias, &state.db)
+            .await;
+
+        logs.push(WebsiteInfo {
+            url: website.url,
+            alias: website.alias,
+            data,
+        });
+    }
+
+    WebsiteLogs { logs }
+}
+```
+
+Once done, we’ll want to create a dynamic route for getting more information about a monitored URL. We can use this page to display things like past incidents/alerts. This function will follow most of the previous handler function except we’re fetching one URL and additionally grabbing any records of HTTP requests that didn’t return `200 OK` and labelling them as “Incidents”:
+
+```rust
+async fn get_website_by_id(
+    State(state): State<AppState>,
+    Path(alias): Path<String>,
+) -> impl AskamaIntoResponse {
+    let website = sqlx::query_as::<_, Website>("SELECT url, alias FROM websites WHERE alias = $1")
+        .bind(&alias)
+        .fetch_one(&state.db)
+        .await
+        .unwrap();
+
+    // .. rest of your code
+}
+```
+
+Then we can write the previous code for getting the uptime percentages per hour, as well as grabbing a vector of `Incident`s:
+
+```rust
+// main.rs
+let last_24_hours_data = get_website_uptime_stats(SplitBy::Hour, &website.alias, &state.db).await;
+let monthly_data = get_website_uptime_stats(SplitBy::Day, &website.alias, &state.db).await;
+
+let incidents = sqlx::query_as::<_, Incident>(
+        "SELECT created_at as time, status from logs where website_alias = $1
+        and status != 200",
+    )
+    .bind(&alias)
+    .fetch_all(&state.db)
+    .await
+    .unwrap();
+
+    let log = WebsiteInfo {
+        url: website.url,
+        alias,
+        data,
+    };
+
+SingleWebsiteLogs { log, incidents, monthly_data }
```

Comment:
Wait, nothing uses `last_24_hours_data`. I don't see where `data` for the log comes from. And `SingleWebsiteLogs` does not have a field for `monthly_data` :sweat_smile:

Reply:
Do we have a step to always dry-run articles before publishing them?

Reply:
I dry-ran this before and it seemed fine. I think there's just a code mismatch, I'll have a look